Adding the initial Tooltip

Rajinsharwar opened the pull request:
<!--
Hi there! Thanks for contributing to WordPress!

Pull Requests in this GitHub repository **must** be linked to a ticket in the WordPress Core Trac instance (https://core.trac.wordpress.org), and are only used for code review. **No pull requests will be merged on GitHub.**

See the WordPress Handbook page on using PRs for Code Review more information: https://make.wordpress.org/core/handbook/contribute/git/github-pull-requests-for-code-review/

If this is your first time contributing, you may also find reviewing these guides first to be helpful:
- FAQs for New Contributors: https://make.wordpress.org/core/handbook/tutorials/faq-for-new-contributors/
- Contributing with Code Guide: https://make.wordpress.org/core/handbook/contribute/
- WordPress Coding Standards: https://make.wordpress.org/core/handbook/best-practices/coding-standards/
- Inline Documentation Standards: https://make.wordpress.org/core/handbook/best-practices/inline-documentation-standards/
- Browser Support Policies: https://make.wordpress.org/core/handbook/best-practices/browser-support/
- Proper spelling and grammar related best practices: https://make.wordpress.org/core/handbook/best-practices/spelling/
-->

PR for adding the initial draft for the Tooltip functionality.

Trac ticket: https://core.trac.wordpress.org/ticket/51006

---
**This Pull Request is for code review only. Please keep all other discussion in the Trac ticket. Do not merge this Pull Request. See [GitHub Pull Requests for Code Review](https://make.wordpress.org/core/handbook/contribute/git/github-pull-requests-for-code-review/) in the Core Handbook for more details.**

Rajinsharwar committed:
Adding the initial Tooltip

Rajinsharwar committed:
Fixing PHPCS

Rajinsharwar committed:
Fixing coding issues - 2

Rajinsharwar committed:
Making compitable with ES5

ayazwpseo commented:
I find myself grappling with precisely the same challenge at the moment. Your guidance is greatly appreciated.
{{{
#!html
<a href="https://casinonewsupdate.com/"></a>
<a href="https://casino-worlds.com/"></a>
<a href="https://casinonewsspot.com/"></a>
<a href="https://casino-hotspot.com/"></a>
<a href="https://casinoworldnews4u.com/"></a>
<a href="https://allaroundcasino.com/"></a>
<a href="https://casinodada.com/"></a>
<a href="https://casinoguide4u.com/"></a>
<a href="http://casinomedia.co.uk/"></a>
<a href="http://casinopost.co.uk/"></a>
<a href="http://casino-days.co.uk/"></a>
<a href="http://casino-guru.co.uk/"></a>
<a href="http://betting-post.com/"></a>
<a href="http://jammymonkey.uk/"></a>
<a href="https://vefeed.com/"></a>
<a href="https://puffmi.com.au"></a>
}}}

sarinajor commented:
<a href='https://capmistoi-6221.weebly.com'>.</a>
<a href='https://capmistoi-6222.weebly.com'>.</a>
<a href='https://capmistoi-6223.weebly.com'>.</a>
<a href='https://capmistoi-6224.weebly.com'>.</a>
<a href='https://capmistoi-6225.weebly.com'>.</a>
<a href='https://capmistoi-6226.weebly.com'>.</a>
<a href='https://capmistoi-6227.weebly.com'>.</a>
<a href='https://capmistoi-6228.weebly.com'>.</a>
<a href='https://capmistoi-6229.weebly.com'>.</a>
<a href='https://capmistoi-6230.weebly.com'>.</a>
<a href='https://capmistoi-6231.weebly.com'>.</a>
<a href='https://capmistoi-6232.weebly.com'>.</a>
<a href='https://capmistoi-6233.weebly.com'>.</a>
<a href='https://capmistoi-6234.weebly.com'>.</a>
<a href='https://capmistoi-6235.weebly.com'>.</a>
<a href='https://capmistoi-6236.weebly.com'>.</a>
<a href='https://capmistoi-6237.weebly.com'>.</a>
<a href='https://capmistoi-6238.weebly.com'>.</a>
<a href='https://capmistoi-6239.weebly.com'>.</a>
<a href='https://capmistoi-6240.weebly.com'>.</a>

Rajinsharwar committed:
Merge branch 'WordPress:trunk' into 51006

Rajinsharwar committed:
Refreshing against trunk, and made improvements for loading and build

Rajinsharwar committed:
Fixed issues with accessibility and design.

Rajinsharwar committed:
Added new arguments and improvements

Rajinsharwar committed:
JShint fix

Rajinsharwar commented:
1. Added validations and _doing_it_wrong fro the arguments passed.
2. Added a new filter that can be used to change the Tooltip icon image. This is similar to the one we have in add_menu() function. We can now pass the URL, or the SVG, or the dashicon class, or even the "none" to style using CSS.

Example usage:
```
add_filter( 'wp_tooltip_icon', 'my_filter', 10, 2 );

function my_filter( $icon_url, $field_id ) {
    $icon_url = "data:image/svg+xml;base64,PHN2ZyB3aWR0aD0iMjAiIGhlaWdodD0iMjAiIHZpZXdib3g9IjAgMCAyMCAyMCIgeG1sbnM9Imh0dHA6Ly93d3cudzMub3JnLzIwMDAvc3ZnIj48cGF0aCBmaWxsPSIjYTdhYWFkIiBkPSJNMTcuNiA4LjVoLTcuNXYzaDQuNGMtLjQgMi4xLTIuMyAzLjUtNC40IDMuNC0yLjYtLjEtNC42LTIuMS00LjctNC43LS4xLTIuNyAyLTUgNC43LTUuMSAxLjEgMCAyLjIuNCAzLjEgMS4ybDIuMy0yLjJDMTQuMSAyLjcgMTIuMSAyIDEwLjIgMmMtNC40IDAtOCAzLjYtOCA4czMuNiA4IDggOGM0LjYgMCA3LjctMy4yIDcuNy03LjgtLjEtLjYtLjEtMS4xLS4zLTEuN3oiIGZpbGxydWxlPSJldmVub2RkIiBjbGlwcnVsZT0iZXZlbm9kZCI+PC9wYXRoPjwvc3ZnPg==";

    return $icon_url;
}
```
This will show an SVG now in the tooltip.

Example:

![image](https://github.com/user-attachments/assets/2a6783ed-ab70-4c4a-bfe4-9c7aaa7f8f12)

3. Added a new action hook that's fired just before the tooltip is rendered in HTML.
4. Loading the styles and scripts for the tooltip before it's rendered to not show a slightly broken tooltip.

Screenshots:

![image](https://github.com/user-attachments/assets/3a476567-85c0-4647-abd0-c8b10a62d124)

Using Tab navigation: 

![image](https://github.com/user-attachments/assets/0733888d-59ac-4f96-abe7-c04500a4df4b)

Rajinsharwar committed:
Adding new filter and action, and also validations

Rajinsharwar committed:
phpcs fix

Rajinsharwar committed:
phpcs fix - 2

Rajinsharwar committed:
phpcs - 3

Rajinsharwar committed:
phpcs - 3

Rajinsharwar committed:
phpcs - 4

Rajinsharwar committed:
phpcs fix

joedolson committed:
Merge branch 'trunk' into pr/5676

joedolson committed:
Restore tooltip function

joedolson committed:
Grunt registry stated `lib`, actual directory is `wp`

joedolson commented on src/js/_enqueues/wp/wp-tooltip.js:
```diff
@@ -0,0 +1,124 @@
+		tooltipButton.addEventListener( 'touchstart', showTooltip );
+		tooltipButton.addEventListener( 'touchend', hideTooltip );
+
+		tooltipButton.style.cursor = 'help';
```
Let's not customize the cursor. Let it use the standard cursor for buttons in WordPress.

joedolson commented on src/wp-includes/css/wp-tooltip.css:
```diff
@@ -0,0 +1,102 @@
+	display: inline-block;
+}
+
+.wp-tooltip-button:focus {
```
This should inherit existing styles for buttons, so that :focus states are standardized. Probably add the class 'button' to the trigger and only add additional styles if there is something that needs to be overwritten.

joedolson commented on src/wp-includes/functions.php:
```diff
@@ -9090,3 +9090,139 @@ function wp_is_heic_image_mime_type( $mime_type ) {
+
+	if ( str_starts_with( $icon_url, 'dashicons-' ) ) {
+		$icon = 'class="dashicons ' . sanitize_html_class( $icon_url ) . '"';
+		$icon_html = '<span ' . $icon . '></span>';
```
Add `aria-hidden` to the image container class, to prevent it from being read by screen readers.

joedolson commented on src/wp-includes/functions.php:
```diff
@@ -9090,3 +9090,139 @@ function wp_is_heic_image_mime_type( $mime_type ) {
+		$icon_html = '<span ' . $icon . '></span>';
+	} elseif ( str_starts_with( $icon_url, 'data:image/svg+xml;base64,' ) ) {
+		$icon = ' style="background-image:url(\'' . esc_attr( $icon_url ) . '\')"';
+		$icon_html = '<svg height = "20" width = "20"' . $icon . '</svg>';
```
Add `aria-hidden` to the SVG, as well as `focusable="false"` to prevent older browser from focusing the SVG independently from the button.

joedolson commented on src/wp-includes/functions.php:
```diff
@@ -9090,3 +9090,139 @@ function wp_is_heic_image_mime_type( $mime_type ) {
+			'6.7.0'
+		);
+
+		$icon_html = '<span class="dashicons dashicons-editor-help"></span>';
```
Add `aria-hidden` here, as well.

joedolson reviewed (changes requested):
This is a first round of feedback; I've still got some more things to look at.

joedolson commented on src/wp-includes/functions.php:
```diff
@@ -9090,3 +9090,139 @@ function wp_is_heic_image_mime_type( $mime_type ) {
+ *
+ * @param string $field_id             The unique identifier for the tooltip container.
+ * @param string $tooltip_text         The text content to be displayed in the tooltip.
+ * @param string $tooltip_button_label Optional. The label for the tooltip button. Default is 'Help'.
```
Change the default value to an empty string, and set the default as `__( 'Help' )` inside the function, so that the default value is translatable.

joedolson commented on src/wp-includes/functions.php:
```diff
@@ -9090,3 +9090,139 @@ function wp_is_heic_image_mime_type( $mime_type ) {
+ * @param string $position             Optional. The default position of the tooltip. Default is 'right'.
+ * 									   Accepted values: 'right', 'left', 'up' and 'down'.
+ */
+function add_tooltip( $field_id, $tooltip_text, $tooltip_button_label = 'Help', $position = 'right' ) {
```
Let's make the arguments into an array, so that the order isn't important. Somebody should be able to set the position without needing to define the button label.

joedolson commented on src/wp-includes/functions.php:
```diff
@@ -9090,3 +9090,139 @@ function wp_is_heic_image_mime_type( $mime_type ) {
+		<button type="button" class="wp-tooltip-button" aria-describedby="<?php echo esc_attr( $field_id ); ?>-tooltip" aria-label="<?php echo esc_attr( $tooltip_button_label ); ?>">
+			<span class="wp-tooltip-button-span"><?php echo $icon_html; ?></span>
+		</button>
+		<div id="<?php echo esc_attr( $field_id ); ?>-tooltip" class="wp-tooltip-content">
```
Add `role="tooltip"`

joedolson commented on src/js/_enqueues/wp/wp-tooltip.js:
```diff
@@ -0,0 +1,124 @@
+
+		// Event listeners for mouse and touch events
+		tooltipContainer.addEventListener( 'mouseenter', showTooltip );
+		tooltipContainer.addEventListener( 'mouseleave', hideTooltip );
```
For accessibility, the tooltip should not be hidden on `mouseleave`; only when either the trigger loses focus or when `esc` is hit.

joedolson reviewed (changes requested):
Some additional comments.

joedolson committed:
Add role=tooltip

joedolson committed:
Don't hide tooltip on mouse leave

joedolson committed:
Don't customize the mouse cursor

joedolson committed:
Inherit styles from existing button classes.

joedolson committed:
Set default string label in a translatable way.

joedolson committed:
Rewrite arguments to pass most args as array.

joedolson committed:
Fix image attribute formatting & add aria-hidden

joedolson committed:
Add focusable=false to svg

joedolson committed:
Add a default field ID.

joedolson committed:
Change keyboard handling

This updates the keyboard handling to match the US Design System's specs for an accessible tooltip. Not having the interaction available on focus reduces discoverability.

joedolson committed:
Support tooltips using JS declaratively

Allows for usage on existing code without requiring the PHP function.

joedolson committed:
Update CSS to remove centered text and modify text reflowing, padding, and size.

joedolson committed:
Fix SVG wrapper; remove whitespace in output

The whitespace in output cause white-space: pre-line to break an extra space.

joedolson committed:
Define undefined vars

joedolson committed:
Simplify declarative use

Only require the wrapper class on any button or a element.

joedolson committed:
Misc. code style changes; update @since

joedolson committed:
Increase button size to 26x26

To meet WCAG 2.5.8

joedolson committed:
Couple version numbers overlooked

swissspidy commented on src/wp-includes/functions.php:
```diff
@@ -5,6 +5,8 @@
  * @package WordPress
  */
 
+use function PHPSTORM_META\map;
```
This needs to be reverted

swissspidy commented on src/wp-includes/functions.php:
```diff
@@ -9090,3 +9092,151 @@ function wp_is_heic_image_mime_type( $mime_type ) {
+			__FUNCTION__,
+			sprintf(
+				/* translators: %s: The invalid argument passed to add_tooltip(). */
+				__( 'The "%s" argument must be a non-empty string.' ),
```
Technically only `$tooltip_text` is checked to be non-empty, so the error message is incorrect for the other variables.

I also find the switch here very confusing. If you fix `$field_id` then you get the next warning for the next variable. That doesn't make sense to me.

I'd just add three checks in a row or so:

```
if ( ! is_string( $field_id ) || '' === $field_id ) {
	_doing_it_wrong(...);
}

if ( ! is_string( $tooltip_text ) || '' === $tooltip_text ) {
	_doing_it_wrong(...);
}

if ( ! is_string( $button_label ) ) {
	_doing_it_wrong(...);
}
```

swissspidy commented on src/wp-includes/functions.php:
```diff
@@ -9090,3 +9092,151 @@ function wp_is_heic_image_mime_type( $mime_type ) {
+	$valid_positions = array( 'top', 'right', 'bottom', 'left' );
+
+	// Validate position input.
+	if ( ! is_string( $position ) || ! in_array( $position, $valid_positions, true ) ) {
```
The `is_string` check is superfluous

swissspidy commented on src/wp-includes/functions.php:
```diff
@@ -9090,3 +9092,151 @@ function wp_is_heic_image_mime_type( $mime_type ) {
+	} elseif ( 'none' === $icon_url ) {
+		$icon_html = '';
+	} else {
+		_doing_it_wrong(
+			__FUNCTION__,
+			sprintf(
+				/* translators: %s: Invalid $icon_html passed in 'pre_wp_tooltip_rendered' hook. */
+				__( 'The passed "%1$s" argument in the "%2$s" filter is invalid.' ),
+				'$icon_html',
+				'pre_wp_tooltip_rendered'
+			),
+			'6.8.0'
```
What does this have to do with the hook below?

Also it's not a filter, it's an action.

swissspidy commented on src/wp-includes/functions.php:
```diff
@@ -9090,3 +9092,151 @@ function wp_is_heic_image_mime_type( $mime_type ) {
+	 *                             e.g. 'dashicons-editor-help'.
+	 *                             Pass 'none' to leave span.wp-tooltip-button-span empty so an icon can be added via CSS.
+	 */
+	add_action( 'pre_wp_tooltip_rendered', $field_id, $tooltip_text, $button_label, $position, $icon_url );
```
What is this good for? `pre_` actions are not so common, and `pre_` filters would be used to short-circuit rendering.

swissspidy commented on src/wp-includes/script-loader.php:
```diff
@@ -848,6 +848,8 @@ function wp_default_scripts( $scripts ) {
 	$scripts->add( 'wp-auth-check', "/wp-includes/js/wp-auth-check$suffix.js", array( 'heartbeat' ), false, 1 );
 	$scripts->set_translations( 'wp-auth-check' );
 
+	$scripts->add( 'wp-tooltip', "/wp-includes/js/wp-tooltip$suffix.js", array( 'jquery' ), false, 1 );
```
Why does this use jQuery?
